### DISCUSSION TITLE
parallelize image optimization

### DISCUSSION THREAD

**Discussion by jemand771** [withastro/roadmap#1008] [Proposal]
### Body

# Summary

Using `<Image />`, `<Picture />` or `getImage()` will cause all images to be optimized sequentially. This is slow when you have a lot of somewhat big source images. I'd like a new option to run this process to run in parallel, i.e. on multiple CPU cores.

# Background & Motivation

I've only tried this with `getImage()` and made a trivial attempt to trick astro into parallelizing it which sadly didn't work:

```ts
inst.allImages = await Promise.all(inst.allImages.map((getter: any) => new Promise(resolve => resolve(getImage({src: getter(), format: "webp"})))))
```

Ideally, I shouldn't even have to manually handle the parallelization logic. Since astro knows what images to optimize anyway, it could handle the "scheduling" of tasks itself, even when my frontmatter/server code is blocking/sequential.

# Goals

* faster builds (especially in CI where there may not be a cache available)
  * more time to develop with astro :sunglasses:
* better utilization of available resources

# Example

I really don't care about what this would look like. I could imagine something like:
```ts
import { defineConfig } from 'astro/config'

export default defineConfig({
  image: {
    workers: 16, // defaults to 1 for backwards compatibility
  },
});
```

An auto-detection option to set workers=number of host CPU cores would also be a nice addition

**Comment by Princesseuh** [comment]
It does not use workers, but image generation is already done concurrently, it processes a few images at once.

**Reply by jemand771** [reply]
I'm not sure it does in my case 🤔 
I have a build with ~500 images. Looking at a recent jenkins job, optimization started at 22:35:28 and finished at 22:51:58:
> 22:51:58 ✓ Completed in 989.71s.

when I sum up the "+123ms" / "+4.5s" times behind each image status line, I get a very similar number: 986 seconds.
I'd expect the sum (cpu time) to be much greater than the actual (wall) time if this was running concurrently

**Reply by Princesseuh** [reply]
Here's the code where this happens: https://github.com/withastro/astro/blob/d1bd1a11f7aca4d2141d1c4665f2db0440393d03/packages/astro/src/core/build/generate.ts#L149-L156

This uses https://github.com/sindresorhus/p-queue to queue the promises

**Reply by delucis** [reply]
I noticed this the other day. The processing is still single-threaded although concurrent — could imagine a multi-thread approach would be faster when available, but it is extra complexity.

**Reply by jemand771** [reply]
aw, sorry, my bad - I mistook concurrency for parallelism 😅 

either way, I guess the feature suggestion stands then: parallel image optimization on multiple threads
